compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -1924,15 +1895,54 @@ object Types {
         case _ => withPrefix(prefix)
       }
 
-    /** Create a NamedType of the same kind as this type, but with a new prefix.
-     */
-    def withPrefix(prefix: Type)(implicit ctx: Context): NamedType
+    /** A reference like this one, but with the given symbol, if it exists */
+    final def withSym(sym: Symbol)(implicit ctx: Context): ThisType =
+      if ((designator ne sym) && sym.exists) NamedType(prefix, sym).asInstanceOf[ThisType]
+      else this
 
-    /** Create a NamedType of the same kind as this type, but with a new namespace.
-     */
-    def withNameSpace(nameSpace: NameSpace)(implicit ctx: Context): NamedType =
-      if (nameSpace == this.nameSpace) this
-      else NamedType(prefix, designator.withNameSpace(nameSpace))
+    /** A reference like this one, but with the given denotation, if it exists */
+    final def withDenot(denot: Denotation)(implicit ctx: Context): ThisType =
+      if (denot.exists) {
+        val adapted = withSym(denot.symbol)
+        if (adapted ne this) adapted.withDenot(denot).asInstanceOf[ThisType]
+        else {
+          setDenot(denot)
```

Comment:
Given the name of this method and its documentation, it's surprising that it can mutate `this`.

Reply:
Good point. I'll update the docs.